### PR TITLE
[stdlib] Fix issue with UTF16 index(_:offsetBy:limitedBy)

### PR DESCRIPTION
**Pull request by airspeedswift**
When the string is empty, the code unit distance to the limit can be zero.

Manifests in this failing:

```
let empty = ""
// should be nil but instead traps
Range(NSRange(location: 1, length: 0), in: empty)
```

rdar://problem/34551055

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by moiseev** [comment]
This seems to be fixing https://bugs.swift.org/browse/SR-4657.
Can we have a test case that proves that? Otherwise LGTM!

**Review by moiseev — APPROVED** [review]

**Comment by airspeedswift** [comment]
@swift-ci please test and merge

**Comment by airspeedswift** [comment]
@swift-ci please test

File: test/stdlib/StringTraps.swift [hunk 1]
```diff
@@ -104,7 +104,36 @@ StringTraps.test("UTF16ViewSubscript/endIndex")
   i = s.utf16.index(after: i)
   i = s.utf16.index(after: i)
   expectCrashLater()
-  s.utf16[i]
+  _ = s.utf16[i]
+}
+
+StringTraps.test("UTF16ViewIndex/offsetLimited")
+  .code {
```
Comment (moiseev) [inline test/stdlib/StringTraps.swift]: `.code` is not necessary.

File: test/stdlib/StringTraps.swift [hunk 2]
```diff
@@ -104,7 +104,36 @@ StringTraps.test("UTF16ViewSubscript/endIndex")
...
+  let sa = "foo"
+  let u16a = sa.utf16
+  let s16 = sa + "🤦🏻‍♀️"
+  let u16 = s16.utf16
+
+  let iaBegin = u16a.index(sa.startIndex, offsetBy: 99, limitedBy: sa.endIndex)
+  expectNil(iaBegin)
+  let iaEnd = u16a.index(sa.endIndex, offsetBy: 99, limitedBy: sa.endIndex)
```
Comment (moiseev) [inline test/stdlib/StringTraps.swift]: It would also be interesting to test for non-positive offsets starting from an `endIndex`.

Comment (airspeedswift) [reply test/stdlib/StringTraps.swift]: Oops, good point!

**Comment by tkremenek** [comment]
@swift-ci test source compatibility

**Pull request merged**